pageserver/src/tenant_tasks.rs

```diff
@@ -72,59 +84,95 @@ async fn compaction_loop(tenant_id: ZTenantId) -> anyhow::Result<()> {
 
-            // Break if tenant is not active
-            if tenant_mgr::get_tenant_state(tenant_id) != Some(TenantState::Active) {
-                break Ok(());
-            }
-            // This should not fail. If someone started us, it means that the tenant exists.
-            // And before you remove a tenant, you have to wait until all the associated tasks
-            // exit.
-            let repo = tenant_mgr::get_repository_for_tenant(tenant_id)?;
+            let tenant = tokio::select! {
+                _ = task_mgr::shutdown_watcher() => {
+                    info!("received GC cancellation request");
+                    return;
+                },
+                tenant_wait_result = wait_for_active_tenant(tenant_id, wait_duration) => match tenant_wait_result {
+                    ControlFlow::Break(()) => return,
+                    ControlFlow::Continue(tenant) => tenant,
+                },
+            };
 
             // Run gc
-            let gc_period = repo.get_gc_period();
-            let gc_horizon = repo.get_gc_horizon();
+            let gc_period = tenant.get_gc_period();
+            let gc_horizon = tenant.get_gc_horizon();
             let mut sleep_duration = gc_period;
             if gc_horizon > 0 {
-                if let Err(e) = repo.gc_iteration(None, gc_horizon, repo.get_pitr_interval(), false)
+                if let Err(e) = tenant.gc_iteration(None, gc_horizon, tenant.get_pitr_interval(), false)
                 {
-                    error!("Gc failed, retrying: {}", e);
-                    sleep_duration = Duration::from_secs(2)
+                    error!("Gc failed, retrying: {e:#}");
+                    sleep_duration = wait_duration;
                 }
             }
 
             // Sleep
             tokio::select! {
                 _ = task_mgr::shutdown_watcher() => {
-                    trace!("received cancellation request");
-                    break Ok(());
+                    info!("received GC cancellation request during idling");
+                    break;
                 },
                 _ = tokio::time::sleep(sleep_duration) => {},
             }
         }
     }
     .await;
     TENANT_TASK_EVENTS.with_label_values(&["stop"]).inc();
-    info!(
-        "GC loop stopped. State is {:?}",
-        tenant_mgr::get_tenant_state(tenant_id)
-    );
-    result
+    trace!("GC loop stopped.");
+}
+
+async fn wait_for_active_tenant(
+    tenant_id: ZTenantId,
+    wait: Duration,
+) -> ControlFlow<(), Arc<Tenant>> {
+    let tenant = loop {
+        match tenant_mgr::get_tenant(tenant_id, false) {
+            Ok(tenant) => break tenant,
+            Err(e) => {
+                error!("Failed to get a tenant {tenant_id}: {e:#}");
+                tokio::time::sleep(wait).await;
+            }
+        }
+    };
+
+    let mut tenant_state_updates = tenant.subscribe_for_state_updates();
+    loop {
+        match tenant_state_updates.changed().await {
```

Comment:
Hmm, if update happened before we called this function we will wait forever until next update comes, right? It looks like this is exactly the case when we call tenant.activate() and then start_loops. 

Tried to reproduce this and inserted panics into gc and compaction routines and I do not see backtraces in logs. Can you reproduce this? Maybe I did something wrong

Reply:
> if update happened before we called this function we will wait forever until next update comes, right?


I think not, we would get the latest updated value and ignore the previous ones:

```rust
#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let (sender, mut receiver) = tokio::sync::watch::channel(0_u32);
    let (wake_sender, wake_receiver) = tokio::sync::oneshot::channel();

    let task_1 = tokio::spawn(async move {
        for i in 1..6 {
            sender.send(i).unwrap();
        }
        wake_sender.send(()).unwrap();
    });

    let task_2 = tokio::spawn(async move {
        wake_receiver.await.unwrap();
        loop {
            match receiver.changed().await {
                Ok(()) => println!("!!!!!! {}", *receiver.borrow()),
                Err(e) => {
                    eprintln!("(((((((((((( {e}");
                    break;
                }
            }
        }
    });

    task_1.await.unwrap();

    Ok(())
}
```

this prints 

```
!!!!!! 5
(((((((((((( channel closed
```

Reply:
We would not get the very first value the channel was initialized with, it's not considered as a "change", yet it's fine since the value is "idle".

I can add a check for an existing value, just in case. I think I've had it already, but got confused and remove it.

Reply:
Without your fix with checking the current value gc/compaction doesnt start at all. Not sure what is the reason then. But with fix it started to work. May worth noting this in comments

Reply:
I've put 

> // if the tenant has a proper status already, no need to wait for anything

so might be ok for now, since not sure how to explain this freezing currently.